File: LinkedLists/DLL_Insertion.java
package LinkedLists;

public class DLL_Insertion {
	class Node{
		int data;
		Node prev,next;
		public Node(int data) {
			this.data=data;
			this.prev=null;
			this.next=null;
		}
	}
	public Node head=null;
	public Node tail=null;
	public void inseratpos(int data,int pos) {
		Node newnode=new Node(data);
		if(pos<0) {
			System.out.print("Invalid position");
			return;
		}
		if(pos==0) {
			newnode.next=head;
			if(head!=null) {
				head.prev=newnode;
			}
			head=newnode;
			if(tail==null) {
				tail=newnode;
			}
			return;
		}
		Node current=head;
		int currpos=0;
		while(currpos<pos-1 && current!=null) {
			current=current.next;
			currpos++;
		}
		if(current==null) {
			System.out.print("Invalid position "+pos);
			return;
		}
		newnode.next=current.next;
		newnode.prev=current;
		
		if(current.next!=null) {
			current.next.prev=newnode;
		}
		else {
			tail=newnode;
		}
		current.next=newnode;
	}
	public void display() {
        Node current = head;
        if (head == null) {
            System.out.println("List is empty");
            return;
        }
        System.out.println("Nodes of doubly linked list: ");
        while (current != null) {
            System.out.print(current.data);
            if(current.next!=null) {
            	System.out.print(" => ");
            }
            current = current.next;
        }
        System.out.println();
    }
	public static void main(String[] args) {
        DLL_Insertion l = new DLL_Insertion();
        l.inseratpos(2, 0);
        l.inseratpos(3, 1);
        l.inseratpos(4, 2);
        l.inseratpos(5, 3);
        l.display();
    }

}

File: LinkedLists/Doubly Linked List/DLL_Insertion.py
class Node:
    def __init__(self,data) -> None:
        self.data=data
        self.prev=None
        self.next=None

class DoublyLinkedList:
    def __init__(self) -> None:
        self.head=None
        self.tail=None
        self.length=0
    
    def addnewnode(self,data,pos):
        newnode=Node(data)
        if pos<0 or pos>self.length:
            print("Invalid Position")
            return
        if pos==0:
            if self.head is not None:
                self.head.prev=newnode
            self.head=newnode
            if self.tail is None:
                self.tail=newnode
        else:
            current=self.head
            for _ in range(pos-1):
                current=current.next
            newnode.prev=current
            newnode.next=current.next
            if current.next is not None:
                current.next.prev=newnode
            else:
                self.tail=newnode
            current.next=newnode
        self.length+=1
    
    def node_beg(self,data):
        newnode=Node(data)
        if self.head is None:
            self.head=newnode
            self.tail=newnode
        else:
            newnode.next=self.head
            self.head.prev=newnode
            self.head=newnode
        
    def node_end(self,data):
        newnode=Node(data)
        if self.head is None:
            self.tail=newnode
            self.head=newnode
        else:
            self.tail.next=newnode
            newnode.prev=self.tail
            self.tail=newnode

    def display(self):
        current=self.head
        if self.head is None:
            print("LinkedList is empty: ")
            return
        print("Nodes of Doubly Linkedlist are: ")
        while current is not None:
            print(current.data,end=" ")
            if current.next is not None:
                print("<=>",end=" ")
            current=current.next
        print()     

    def __len__(self):
        return self.length
        


if __name__=="__main__":
    l=DoublyLinkedList()
    l.addnewnode(2,0)
    l.addnewnode(3,1)
    l.addnewnode(4,2)
    l.display()
    l.addnewnode("M",len(l))
    l.node_end("last")
    l.node_beg(1)
    l.display()
                 

'''
Nodes of Doubly Linkedlist are: 
2 <=> 3 <=> 4
Nodes of Doubly Linkedlist are:
1 <=> 2 <=> 3 <=> 4 <=> M <=> last
'''
